sympy/solvers/inequalities.py

```diff
@@ -560,21 +543,40 @@ def valid(x):
                         'relationship did not evaluate: %s' % r)
 
             singularities = []
-            for d in denoms(expr, gen):
-                singularities.extend(solvify(d, gen, domain))
-            if not continuous:
-                domain = continuous_domain(expanded_e, gen, domain)
 
             include_x = '=' in expr.rel_op and expr.rel_op != '!='
+            try:
+                for d in denoms(expr, gen):
+                    singularities.extend(solvify(d, gen, domain))
+                if not continuous:
+                    domain = continuous_domain(expanded_e, gen, domain)
+                if gen not in n.free_symbols and len(e.free_symbols) > 1:
+                    raise ValueError
+                # this might raise ValueError on its own
+                # or it might give None...
+                solns = solvify(e, gen, domain)
+                if solns is None:
+                    # in which case we raise ValueError
+                    raise ValueError
+                else:
+                    discontinuities = set(domain.boundary -
+                        FiniteSet(domain.inf, domain.sup))
+                    # remove points that are not between inf and sup of domain
+                    critical_points = FiniteSet(*(solns + singularities + list(
+                        discontinuities))).intersection(
+                        Interval(domain.inf, domain.sup,
+                        domain.inf not in domain, domain.sup not in domain))
+                    if isinstance(critical_points,Intersection):
+                        raise ValueError
```

Comment:
```suggestion
                        solns = []
```

Reply:
```
                solns = solvify(e, gen, domain)

                if solns:
                    discontinuities = set(domain.boundary -
                        FiniteSet(domain.inf, domain.sup))
                    # remove points that are not between inf and sup of domain
                    critical_points = FiniteSet(*(solns + singularities + list(
                        discontinuities))).intersection(
                        Interval(domain.inf, domain.sup,
                        domain.inf not in domain, domain.sup not in domain))
                    if isinstance(critical_points,Intersection):
                        solns = []
                if solns is None:
                    # in which case we raise ValueError
                    raise ValueError
```
Is this the correct way to proceed ?

Reply:
The remaining code about raising errors was not written by me...
only the below line of code are addition (+30 -28) remaining are just rearrangements...

```
if isinstance(critical_points,Intersection):
           solns = []
```

Reply:
I think it would be better to test `if isinstance(critical_points, FiniteSet)`. That way we test for a case that we know we can handle. Also setting `solns = []` is cryptic. It would be better to test closer to where the test is needed.

Reply:
> I think it would be better to test `if isinstance(critical_points, FiniteSet)`. That way we test for a case that we know we can handle. Also setting `solns = []` is cryptic. It would be better to test closer to where the test is needed.

At this point, we're mostly concerned about not to get `intersection` set (https://github.com/sympy/sympy/pull/20824#issuecomment-786835550)

> Also setting `solns = []` is cryptic

So, what about `solns = None`( if it's intersection then `solns` should be None). Also, the `solns` is not used anywhere in the function except in `critical_points`.

Reply:
@oscarbenjamin 
please give your view on this (https://github.com/sympy/sympy/pull/20824#discussion_r584020497)

Reply:
> > I think it would be better to test `if isinstance(critical_points, FiniteSet)`. That way we test for a case that we know we can handle. Also setting `solns = []` is cryptic. It would be better to test closer to where the test is needed.
> 
> At this point, we're mostly concerned about not to get `intersection` set ([#20824 (comment)](https://github.com/sympy/sympy/pull/20824#issuecomment-786835550))
> 
> > Also setting `solns = []` is cryptic
> 
> So, what about `solns = None`( if it's intersection then `solns` should be None). Also, the `solns` is not used anywhere in the function except in `critical_points`.
> 
> I think the `solns` variable should be renamed to something like `critical_solns` so that it will be more specific ( where using `solns` give a broad sense. )

Or else ,... what about taking

```
if not isinstance(critical_points,FiniteSet):
    solns = None
```

Reply:
> Or else ,... what about taking
> 
> ```
> if not isinstance(critical_points,FiniteSet):
>     solns = None
> ```

I think that would probably be better.

Reply:
> I think the solns variable should be renamed to something like critical_solns so that it will be more specific ( where using solns give a broad sense. )

And, what about this?